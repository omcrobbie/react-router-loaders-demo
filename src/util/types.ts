import { LoaderFunction } from "react-router-dom";

export type ContactType = {
  id: string;
  createdAt: number;
  first: string;
  last: string;
  avatar: string;
  twitter: string;
  notes: string;
  favorite: boolean;
};

export type LoaderData<TLoaderFn extends LoaderFunction> = Awaited<
  ReturnType<TLoaderFn>
> extends Response | infer D
  ? D
  : never;
