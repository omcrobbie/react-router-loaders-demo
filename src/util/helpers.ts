import { SerializeFrom } from "@remix-run/node";
import { LoaderFunction, useAsyncValue } from "react-router-dom";
export { useLoaderData as useTypedLoaderData } from "@remix-run/react";

export const useTypedAsyncValue = <
  T extends LoaderFunction,
  K extends keyof SerializeFrom<T>
>() => {
  return useAsyncValue() as Awaited<SerializeFrom<T>[K]>;
};
