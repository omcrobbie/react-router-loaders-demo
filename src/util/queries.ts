import { queryOptions } from "@tanstack/react-query";
import { getContact, getContacts } from "../contacts";

export const contactsQueryOptions = (q: string | null) =>
  queryOptions({
    queryKey: ["contacts", q],
    queryFn: () => getContacts(q),
  });

export const contactQueryOptions = (contactId: string) =>
  queryOptions({
    queryKey: ["contact", contactId],
    queryFn: async () => {
      const contact = await getContact(contactId);
      if (!contact)
        throw new Response("", {
          status: 404,
          statusText: "Not Found",
        });
      return contact;
    },
  });
