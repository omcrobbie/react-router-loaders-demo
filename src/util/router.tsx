import { createRoutesFromElements, Route } from "react-router-dom";
import ErrorPage from "../error-page";
import Index from "../routes";
import Contact from "../routes/contact";
import EditContact from "../routes/edit";
import Root from "../routes/root";
import {
  contactAction,
  contactLoader,
  contactsLoader,
  deleteAction,
  editAction,
  favoriteAction,
} from "./functions";

export const routes = createRoutesFromElements(
  <Route
    path="/"
    element={<Root />}
    errorElement={<ErrorPage />}
    loader={contactsLoader}
    action={contactAction}
  >
    <Route errorElement={<ErrorPage />}>
      <Route index element={<Index />} />
      <Route
        errorElement={<ErrorPage />}
        path="/contacts/:contactId"
        element={<Contact />}
        loader={contactLoader}
        action={favoriteAction}
      />
      <Route
        path="/contacts/:contactId/edit"
        element={<EditContact />}
        action={editAction}
        loader={contactLoader}
      />
      <Route
        path="/contacts/:contactId/destroy"
        action={deleteAction}
        errorElement={<div>Oops! There was an error.</div>}
      />
    </Route>
  </Route>
);
