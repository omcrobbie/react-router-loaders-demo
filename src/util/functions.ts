import { QueryClient } from "@tanstack/react-query";
import { ActionFunction, LoaderFunction, redirect } from "react-router-dom";
import { createContact, deleteContact, updateContact } from "../contacts";
import { deferredLoader } from "./helpers";
import { contactQueryOptions, contactsQueryOptions } from "./queries";

export const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      staleTime: 1000 * 60 * 5,
    },
  },
});

export const contactLoader = (({ params }) => {
  return queryClient.ensureQueryData(contactQueryOptions(params.contactId!));
}) satisfies LoaderFunction;

export const contactsLoader = deferredLoader(({ request }) => {
  const url = new URL(request.url);
  const q = url.searchParams.get("q");
  const options = contactsQueryOptions(q);
  return {
    contacts: queryClient.ensureQueryData(options),
  };
});

export const contactAction: ActionFunction = async () => {
  await queryClient.invalidateQueries({ queryKey: ["contacts"] });
  const contact = await createContact();
  return redirect(`/contacts/${contact.id}/edit`);
};

export const editAction: ActionFunction = async ({ request, params }) => {
  const formData = await request.formData();
  const updates = Object.fromEntries(formData);
  await updateContact(params.contactId!, updates);
  return redirect(`/contacts/${params.contactId}`);
};

export const favoriteAction: ActionFunction = async ({ request, params }) => {
  const formData = await request.formData();
  return updateContact(params.contactId!, {
    favorite: formData.get("favorite") === "true",
  });
};

export const deleteAction: ActionFunction = async ({ params }) => {
  await deleteContact(params.contactId!);
  return redirect("/");
};
