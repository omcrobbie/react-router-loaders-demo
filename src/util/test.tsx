import { render } from "@testing-library/react";
import { createMemoryRouter } from "react-router-dom";
import { App } from "../App";
import { routes } from "./router";

export const renderApp = (opts?: { url: string }) => {
  const testRouter = createMemoryRouter(routes, {
    initialEntries: [opts?.url ?? "/"],
  });
  return render(<App router={testRouter} />);
};
