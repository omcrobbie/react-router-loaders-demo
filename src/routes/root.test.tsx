import { screen } from "@testing-library/react";
import { expect, test, vi } from "vitest";
import { debug } from "vitest-preview";
import { renderApp } from "../util/test";

vi.mock("localforage");
test("root - renders list of contacts", async () => {
  const { container } = renderApp();
  await screen.findByTestId("contacts-list");
  debug();
  expect(container).toMatchSnapshot();
});
