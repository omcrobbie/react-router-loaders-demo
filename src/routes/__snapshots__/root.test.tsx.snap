// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`root - renders list of contacts 1`] = `
<div>
  <div
    id="sidebar"
  >
    <h1>
      React Router Contacts
    </h1>
    <div>
      <form
        action="/"
        id="search-form"
        method="get"
        role="search"
      >
        <input
          aria-label="Search contacts"
          class=""
          id="q"
          name="q"
          placeholder="Search"
          type="search"
          value=""
        />
        <div
          aria-hidden="true"
          hidden=""
          id="search-spinner"
        />
        <div
          aria-live="polite"
          class="sr-only"
        />
      </form>
      <form
        action="/"
        method="post"
      >
        <button
          type="submit"
        >
          New
        </button>
      </form>
    </div>
    <nav
      data-testid="contacts-list"
    >
      <ul>
        <li>
          <a
            class=""
            href="/contacts/jb6bezu"
          >
            Arlo
             
            Collins
             
          </a>
        </li>
        <li>
          <a
            class=""
            href="/contacts/tmfta2f"
          >
            Stuart
             
            Dietrich
             
          </a>
        </li>
        <li>
          <a
            class=""
            href="/contacts/verp7yw"
          >
            Ivy
             
            Faheys
             
          </a>
        </li>
        <li>
          <a
            class=""
            href="/contacts/fneyyep"
          >
            Milo
             
            Hagene
             
          </a>
        </li>
        <li>
          <a
            class=""
            href="/contacts/ivtdycc"
          >
            Maeve
             
            Keeling
             
          </a>
        </li>
        <li>
          <a
            class=""
            href="/contacts/tjhmw2g"
          >
            Demarcus
             
            Kub
             
            <span>
              ★
            </span>
          </a>
        </li>
        <li>
          <a
            class=""
            href="/contacts/kt1pa1k"
          >
            Kaitlyn
             
            Leuschke
             
          </a>
        </li>
        <li>
          <a
            class=""
            href="/contacts/z3te9nl"
          >
            Lorenzo
             
            Lowe
             
          </a>
        </li>
        <li>
          <a
            class=""
            href="/contacts/5kdlev3"
          >
            Abdullah
             
            Veum-O'Conners
             
          </a>
        </li>
        <li>
          <a
            class=""
            href="/contacts/5jpv6le"
          >
            Reggie
             
            Veums
             
            <span>
              ★
            </span>
          </a>
        </li>
        <li>
          <a
            class=""
            href="/contacts/vc9pp09"
          >
            Margaret
             
            Zboncakz
             
          </a>
        </li>
      </ul>
    </nav>
  </div>
  <div
    class=""
    id="detail"
  >
    <p
      id="zero-state"
    >
      This is a demo for React Router.
      <br />
      Check out
       
      <a
        href="https://reactrouter.com"
      >
        the docs at reactrouter.com
      </a>
      .
    </p>
  </div>
</div>
`;
